ephread/inkgd
# Memory leakage partial fix [WIP]

Hello!
As you may know I'm trying to get rid of the memory leakage that are currently present in this repository. Here's an example of memory leakage that I encounter:
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/42484461/84714313-9daedc00-af6d-11ea-9cd5-55919e788d2e.gif)

I am slowly trying to isolate the multiple memory leakages that are present and will push to this branch whenever there are any advancements.

For isolating each memory leak **I remove parts of the '_the_intercept.INK_'-file until I can isolate a particular leakage issue**. As you can see, I had to remove quite a lot... I hope that in the future I can re-add the _raise()_- and _lower()_-functions, that will be the next goal.

### Currently this branch implements the following fixes:
- **Get rid of strong reference to parent**, which created a circular reference between the `_content`-array (as found in the parent Container) and the _parent_-reference (As found in the child Container)
All parent variables have been replaced with weak references to the parent instead.
- **Get rid of the strong container-reference in pointer.gd**, which again leads to circular references as a container can contain a Divert, which then contains a Pointer pointing to the container.

### How to check for memory leakage?

1. Set _verbose_stdout_ to _true_ in the ProjectSettings so you get as much information as possible.
2. Run the '_Inkrunner.tscn_' and wait until the loading screen is gone.
3. Exit the game by clicking the X-button
4. Check the console for any memory leak messages

### Console log for current **PRUNED** '_the_intercept.INK_'-file
Result with strong parent reference:
![leakage](https://user-images.githubusercontent.com/42484461/84714096-f5007c80-af6c-11ea-831e-066949de4bca.PNG)

Result with weak parent reference:
![nothing](https://user-images.githubusercontent.com/42484461/84714100-f631a980-af6c-11ea-9658-4b7364e9ffbe.PNG)

I hope to work on these leakage whenever I can... but it's tricky work and any help is welcome!
